File: tools/dev/launchDev/child-build.ts
import { resolve } from "pathe";
import { build } from "vite";
import packageJson from "../../../package.json" with { type: "json" };

const r = (value: string): string => {
  return resolve(import.meta.dirname!, "../../..", value);
};

async function launchBuild(mode: string, buildid2: string) {
  if (mode.startsWith("dev")) {
    await Promise.all([
      new Deno.Command(Deno.execPath(), {
        args: ["task", "build", `--env.MODE=${mode}`],
        cwd: r("./src/core/glue/startup"),
      }).output(),
      new Deno.Command(Deno.execPath(), {
        args: [
          "task",
          "build",
          `--env.__BUILDID2__=${buildid2}`,
          `--env.__VERSION2__=${packageJson.version}`,
        ],
        cwd: r("./src/core/glue/loader-modules"),
      }).output(),
    ]);
    await Promise.all([
      new Deno.Command(Deno.execPath(), {
        args: ["task", "build", `--env.MODE=${mode}`],
        cwd: r("./src/core/glue/startup"),
      }).output(),
      build({
        root: r("./src/core/glue/loader-features"),
        configFile: r("./src/core/glue/loader-features/vite.config.ts"),
        define: {
          "import.meta.env.__BUILDID2__": `"${buildid2 ?? ""}"`,
          "import.meta.env.__VERSION2__": `"${packageJson.version}"`,
        },
        base: "chrome://noraneko/content",
      }),
      new Deno.Command(Deno.execPath(), {
        args: [
          "task",
          "build",
          `--env.__BUILDID2__=${buildid2}`,
          `--env.__VERSION2__=${packageJson.version}`,
        ],
        cwd: r("./src/core/glue/loader-modules"),
      }).output(),
    ]);
  }
}

{
  //* main
  await launchBuild(Deno.args[0], Deno.args[1]);
}

File: tools/misc/launchDev/child-build.rb
require 'json'

def abs(path)
  File.expand_path(File.join('..', '..', '..', path), __dir__)
end

def package_version
  JSON.parse(File.read(abs('package.json')))['version']
end

def run(cmd, dir)
  system(*cmd, chdir: dir)
end

def launch_build(mode, buildid2)
  return unless mode.start_with?('dev')
  run(['deno', 'task', 'build', "--env.MODE=#{mode}"], abs('src/core/glue/startup'))
  run(['deno', 'task', 'build', "--env.__BUILDID2__=#{buildid2}", "--env.__VERSION2__=#{package_version}"], abs('src/core/glue/loader-modules'))
  run(['deno', 'task', 'build', "--env.MODE=#{mode}"], abs('src/core/glue/startup'))
  run(['deno', 'task', 'build', "--env.__BUILDID2__=#{buildid2}", "--env.__VERSION2__=#{package_version}"], abs('src/core/glue/loader-modules'))
end

if $PROGRAM_NAME == __FILE__
  launch_build(ARGV[0] || 'dev', ARGV[1] || 'default-build-id')
end